src/base/memory_monitor_sr.cpp

```diff
@@ -281,61 +302,103 @@ namespace cubperf
+		subcomp_idx = m_component[comp_idx]->get_subcomp_index (info[idx].subcomp_name);
 		if (subcomp_idx != -1)
 		  {
 		    subcomp_skip = true;
 		  }
 
 		if (!subcomp_skip)
 		  {
-		    subcomp_idx = m_component[comp_idx]->add_subcomponent (info[cnt].subcomp_name);
+		    subcomp_idx = m_component[comp_idx]->add_subcomponent (info[idx].subcomp_name);
 		  }
-		/* XXX: add error if subcomp_idx > MAX_SUBCOMP_IDX */
+		assert (subcomp_idx < MAX_COMP_IDX);
 	      }
 	  }
+	else
+	  {
+	    assert (info[idx].subcomp_name == NULL);
+	  }
 
-	m_comp_idx_map.push_back (make_stat_idx_map (comp_idx, subcomp_idx));
+	m_comp_idx_map.push_back (make_comp_idx_map_val (comp_idx, subcomp_idx));
 
-	cnt++;
+	idx++;
       }
   }
 
   int mmon_module::aggregate_stats (const MMON_MODULE_INFO &info)
   {
-    return 0;
+    int error = NO_ERROR;
+    return error;
   }
 
-  void mmon_module::add_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init, bool expand)
+  void mmon_module::get_comp_and_subcomp_idx (MMON_STAT_ID stat_id, int &comp_idx, int &subcomp_idx)
   {
-    /* description of add_stat(..).
-     * 1) if init == true, add init_stat
-     * 2) then, add cur_stat
-     * 3) compare with peak_stat
-     * 4) if cur_stat(new) > peak_stat, update peak_stat
-     * 5) if expand == true, expand_count++
-     * 6) call component->add_stat(size, subcomp_idx, init, expand) */
-    return;
+    int comp_map_idx = get_comp_map_idx (stat_id);
+    int idx_map_val;
+
+    idx_map_val = m_comp_idx_map[comp_map_idx];
+    comp_idx = get_comp_idx (idx_map_val);
+    subcomp_idx = get_subcomp_idx (idx_map_val);
+  }
+
+  void mmon_module::add_stat (MMON_STAT_ID stat_id, int64_t size)
+  {
+    int comp_idx, subcomp_idx;
+
+    get_comp_and_subcomp_idx (stat_id, comp_idx, subcomp_idx);
+
+    m_stat.cur_stat += size;
+
+    if (m_stat.cur_stat > m_stat.peak_stat)
+      {
+	m_stat.peak_stat = m_stat.cur_stat.load ();
+      }
+
+    /* If there is a component info */
+    if (comp_idx != mmon_module::MAX_COMP_IDX)
+      {
+	assert (comp_idx < m_component.size ());
+	m_component[comp_idx]->add_stat (size, subcomp_idx);
+      }
+  }
+
+  void mmon_module::add_expand_count (MMON_STAT_ID stat_id)
+  {
+    int comp_idx, subcomp_idx;
+
+    m_stat.expand_count++;
+
+    get_comp_and_subcomp_idx (stat_id, comp_idx, subcomp_idx);
+
+    /* If there is a component info */
+    if (comp_idx != mmon_module::MAX_COMP_IDX)
+      {
+	assert (comp_idx < m_component.size ());
+	m_component[comp_idx]->add_expand_count ();
+      }
   }
 
-  void mmon_module::sub_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init)
+  void mmon_module::end_init_phase ()
   {
-    /* description of sub_stat(size, init).
-     * 1) if init == true, sub init_stat
-     * 2) then, sub cur_stat
-     * 3) call component->sub_stat(size, subcomp_idx, init) */
-    return;
+    m_stat.init_stat = m_stat.cur_stat.load ();
+
+    for (auto i = 0; i < m_component.size (); i++)
```

Comment:
The `for-each` is better if possible. It's open to changing its container and probably faster. 
```suggestion
    for (const &auto comp :  m_component)
```

Reply:
Done.